### PR TITLE
Support add_param_group

### PR DESCRIPTION
**Pull request by mcarilli**
Possible fix for #292

### CODE REVIEW

**Comment by vfdev-5** [comment]
@mcarilli does this PR alternate internal optimizer's `param_groups` ? In Ignite contrib lr schedulers take directly references to groups instead of optimizer. So, once user `amp.initialize` after defining an lr scheduler it wont work as groups seem to be redefined. Any chance to keep the references to groups the same ?

**Comment by mcarilli** [comment]
Amp patches the [original optimizer.add_param_group](https://pytorch.org/docs/stable/_modules/torch/optim/optimizer.html#Optimizer.add_param_group) to mutate the incoming new group in place (creating FP32 master params from FP16 model params if necessary), then [calls the original optimizer.add_param_group](https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L405) to add the new group.

When you call `optimizer.add_param_group(new_group)`, it should mutate the incoming `new_group` dict in place.  In other words if you (or some LR scheduler) stash some external reference to `new_group`, then call `optimizer.add_param_group(new_group)`, the external references should still point to the same dict object that ends up being saved in `optimizer.param_groups` after the call to `add_param_group`.  Some of the params owned within the dict may have been changed (for example, some may have been cast from fp16 to fp32).  But saying e.g. `external_reference['lr'] += 1.0` later in the script should be visible/picked up by the optimizer.  Is this not what you're seeing?

**Comment by vfdev-5** [comment]
@mcarilli thanks for the answer. I see more or less what you mean. But I can not say whether it is related to my problem. More precisely my problem is the following :
```python
import torch
print(torch.__version__)

model = torch.nn.Linear(1, 1).to('cuda')
optim = torch.optim.SGD(model.parameters(), lr=0.1)

param_groups = optim.param_groups
param_group = optim.param_groups[0]
print(id(param_groups), id(param_group), param_group['lr'])

from apex import amp
model, optim = amp.initialize(model, optim, opt_level="O2", num_losses=1)

new_param_groups = optim.param_groups
new_param_group = optim.param_groups[0]
print(id(new_param_groups), id(new_param_group), new_param_group['lr'])

# As ignite's classes work directly on param_groups passed from the previous optimizer
# internally it sets for example lr to another value:
param_group['lr'] = 0.01
# but current optimizer is not updated
print(optim.param_groups[0]['lr'])
```
with the output
```
1.1.0
140687541229960 140687957656704 0.1
Selected optimization level O2:  FP16 training with FP32 batchnorm and FP32 master weights.

Defaults for this optimization level are:
enabled                : True
opt_level              : O2
cast_model_type        : torch.float16
patch_torch_functions  : False
keep_batchnorm_fp32    : True
master_weights         : True
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True
opt_level              : O2
cast_model_type        : torch.float16
patch_torch_functions  : False
keep_batchnorm_fp32    : True
master_weights         : True
loss_scale             : dynamic
140689438827528 140687957665472 0.1
0.1
```

As you can see, once I initialized model and optimizer with amp. Optimizer's param_groups are not the same as the previous ones. Any ideas why it is so and whether we could keep them the same ?

**Comment by mcarilli** [comment]
So, you're worried about the effects of the amp.initialize call, not the effects of calling add_param_group explicitly?   The add_param_group PR should not have altered the behavior of amp.initialize.

**Comment by vfdev-5** [comment]
Yes, seems like that. However, I do not see where in `_process_optimizer` function param_groups are altered. Actually, why apex need to alter `param_groups` instead of maybe replacing "params" ?
